File: java/3. Longest Substring Without Repeating Characters/LongestSubstring.java
import java.util.HashMap;

public class LongestSubstring {

    static int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> window = new HashMap<Character, Integer>();
        Integer tempLen = 0, maxLen = 0, start = 0;

        for(int i = 0; i < s.length(); i++) {
            if (window.containsKey(s.charAt(i))) {
                for(int j = start; j < window.get(s.charAt(i)); j++) window.remove(s.charAt(j));

                tempLen += start - window.get(s.charAt(i));
                start = window.get(s.charAt(i)) + 1;
                window.replace(s.charAt(i), i);
            } else {
                tempLen++;
                maxLen = Math.max(maxLen, tempLen);
                window.put(s.charAt(i), i);
            }
        }

        return maxLen;
    }
    public static void main(String[] args) {
        String s = "hello world!";

        System.out.println(lengthOfLongestSubstring(s));
    }
}

File: javascript/3. Longest Substring Without Repeating Characters/longest_substring.js
/**
 * @param {string} s
 * @return {number}
 */
var length_of_longest_substring = function(s) {
    var window = new Map();
    var start = 0;
    var temp_len = 0;
    var max_len = 0;

    for(var i = 0; i < s.length; i++) {
        if(window.has(s[i])) {
            for(var j = start; j<window.get(s[i]); j++) window.delete(s[j]);

            temp_len += start - window.get(s[i]);
            start = window.get(s[i]) + 1;
        } else {
            temp_len += 1;
            max_len = Math.max(max_len, temp_len);
        }
        
        window.set(s[i], i);
    }

    return max_len;
}

var s = "hello world!";
console.log(length_of_longest_substring(s));